### ISSUE TITLE
NaN gradient only under `vmap` for ControlTerm with NaNs in control

### ISSUE THREAD

**Issue by RunnersNum40** [patrick-kidger/diffrax#701]
I have a case where `vmap`ing over gradient functions changes the behaviour compared to individual calls. When using a control term that has NaNs, the gradient is fine as long as the region of integration avoids them. Under `vmap` though, this is no longer true. I expect this is due to a combination of some `cond` being converted to `select` à la [https://docs.jax.dev/en/latest/_autosummary/jax.lax.cond.html](https://docs.jax.dev/en/latest/_autosummary/jax.lax.cond.html) and maybe the thrilling [Gradients contain NaN where using `where`](https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where), but I haven't been able to understand the internals of Diffrax well enough to diagnose it myself. It's probably the case that this can be avoided with some clever guards wrapping it that mask out dangerous inputs, but I've not been able to get it to work that way.

This is the most minimal reproduction I could manage (unless you'd like less typing :):
```py
import equinox as eqx
import jax
from diffrax import (
    ControlTerm,
    CubicInterpolation,
    Euler,
    backward_hermite_coefficients,
    diffeqsolve,
)
from jax import numpy as jnp
from jaxtyping import Array, Float


@jax.grad
def solve(scale: Float[Array, ""], ts: Float[Array, " n"]) -> Float[Array, ""]:
    scale = eqx.debug.backward_nan(scale, terminate=False)
    coeffs = backward_hermite_coefficients(ts, ts)
    term = ControlTerm(lambda *_: scale, CubicInterpolation(ts, coeffs))

    solution = diffeqsolve(
        terms=term,
        solver=Euler(),
        t0=jnp.nanmin(ts),
        t1=jnp.nanmax(ts),
        dt0=1.0,
        y0=jnp.array(0.0),
    )

    return solution.ys[0]  # pyright: ignore


scale = jnp.array(1.0)
cases = jnp.array(
    [
        [jnp.nan, 0.0],
        [0.0, 1.0],
        [0.0, jnp.nan],
    ]
)

print("Gradients:", jnp.array([solve(scale, ts) for ts in cases]))
print("Gradients (vmap):", jax.vmap(solve, in_axes=(None, 0))(scale, cases))
```

The output is:
```
   primals=Array(1., dtype=float32, weak_type=True)
cotangents=Array(0., dtype=float32, weak_type=True)
   primals=Array(1., dtype=float32, weak_type=True)
cotangents=Array(1., dtype=float32, weak_type=True)
   primals=Array(1., dtype=float32, weak_type=True)
cotangents=Array(0., dtype=float32, weak_type=True)
Gradients: [0. 1. 0.]

   primals=Array(1., dtype=float32, weak_type=True)
cotangents=Array(nan, dtype=float32, weak_type=True)
   primals=Array(1., dtype=float32, weak_type=True)
cotangents=Array(1., dtype=float32, weak_type=True)
   primals=Array(1., dtype=float32, weak_type=True)
cotangents=Array(nan, dtype=float32, weak_type=True)
Gradients (vmap): [nan  1. nan]
```

In my actual use case, the `vmap` is inside the grad transformed function, but I think having it outside makes my example clearer since I can show the exact same function working then not working.

**Comment by RunnersNum40** [comment]
I've created a work around that masks out NaN producing inputs. It's possible that this will change the output slightly if there are leading NaNs, by changing the initial slope of the spline. I currently only care about trailing NaNs in my use case and the backwards differences ensures this won't change the output.

```py
def check_make_safe_invariants(
    ts: Float[Array, " n"], ys: Float[Array, " n d"]
) -> tuple[Float[Array, " n"], Float[Array, " n d"]]:
    ts_nan = jnp.isnan(ts)
    ys_row_nan = jnp.all(jnp.isnan(ys), axis=-1)

    ts = eqx.error_if(ts, ts_nan != ys_row_nan, "NaN rows in ys must match NaNs in ts.")

    transitions = jnp.sum(ts_nan[:-1] != ts_nan[1:])
    edges_only = (transitions <= 1) | ((transitions == 2) & ts_nan[0] & ts_nan[-1])
    any_valid = jnp.any(~ts_nan)
    ts = eqx.error_if(
        ts,
        ~(edges_only & any_valid),
        "NaNs in ts may only appear at the edges; the non-NaN region must be a single contiguous block with at least one valid timestamp.",
    )

    increasing = jnp.where(ts_nan[1:] | ts_nan[:-1], True, ts[1:] > ts[:-1])
    ts = eqx.error_if(ts, ~increasing, "ts must be strictly increasing where not NaN.")
    return ts, ys


def make_monotonic_increasing(ts: Float[Array, " n"]) -> Float[Array, " n"]:
    """
    Replace edge NaNs while making ts strictly increasing.

    Assumes make_safe_invariants have been checked.
    """
    n = ts.shape[0]
    idxs = jnp.arange(n)
    nan_mask = jnp.isnan(ts)

    first_valid_idx = jnp.argmax(~nan_mask)
    first_val = ts[first_valid_idx]
    step_left = jnp.maximum(jnp.spacing(first_val), jnp.finfo(ts.dtype).eps)
    ts_filled = jnp.where(
        idxs < first_valid_idx,
        first_val - step_left * (first_valid_idx - idxs),
        ts,
    )

    last_valid_idx = n - 1 - jnp.argmax((~nan_mask)[::-1])
    last_val = ts[last_valid_idx]
    step_right = jnp.maximum(jnp.spacing(last_val), jnp.finfo(ts.dtype).eps)
    ts_filled = jnp.where(
        idxs > last_valid_idx,
        last_val + step_right * (idxs - last_valid_idx),
        ts_filled,
    )

    return ts_filled


def pad_ends(xs: Float[Array, " n d"]) -> Float[Array, " n d"]:
    """
    Replace NaN rows at the start and end of xs with the nearest valid row.

    Assumes make_safe_invariants have been checked.
    """
    n = xs.shape[0]
    idxs = jnp.arange(n)
    row_nan = jnp.all(jnp.isnan(xs), axis=-1)

    first_valid_idx = jnp.argmax(~row_nan)
    leading_mask = (idxs < first_valid_idx) & row_nan
    filled = jnp.where(
        leading_mask[:, None],
        jnp.broadcast_to(xs[first_valid_idx], xs.shape),
        xs,
    )

    last_valid_idx = n - 1 - jnp.argmax((~row_nan)[::-1])
    trailing_mask = (idxs > last_valid_idx) & row_nan
    filled = jnp.where(
        trailing_mask[:, None],
        jnp.broadcast_to(xs[last_valid_idx], xs.shape),
        filled,
    )
    return filled


def make_safe(
    ts: Float[Array, " n"], ys: Float[Array, " n d"]
) -> tuple[Float[Array, " n"], Float[Array, " n d"]]:
    ts, ys = check_make_safe_invariants(ts, ys)
    return make_monotonic_increasing(ts), pad_ends(ys)


def safe_control(
    ts: Float[Array, " n"], ys: Float[Array, " n d"]
) -> CubicInterpolation:
    ts, ys = make_safe(ts, ys)
    coeffs = backward_hermite_coefficients(ts, ys)
    return CubicInterpolation(ts, coeffs)
```

Using the masked values eliminates the problem for me. I'm going to leave this issue open to see if there is further discussion or advice. While it's not a huge problem, it would be nice if diffrax handled issues like this internally, I'd be thrilled if that was something that could be supported.

**Comment by patrick-kidger** [comment]
I've realised that I meant to offer a response/explanation for this one for a while! Apologies for the huge delay :D

So basically what you're observing here is a combination of two things:
1. `vmap`-of-a-loop works by running enough 'fake steps' that all batch elements have completed;
2. `0 * nan = nan` (not zero!)

The former means that the batched case actually performs computations on the nan-containing batch elements (which are then thrown away in the primal, but still present for the backward), whilst the latter means that autodiff will produce nans.

Taking a step back, there's actually a general numerical computing principle here, which is that nans and reverse-mode autodiff do not mix. Once a nan has been produced then unless one is reasoning very carefully about the computation graph, it might well be saved as part of the data needed for the subsequent backward pass. For this reason I generally don't aim to support computations which have nans in the input + I recommend using something like `inf` as a sentinel value instead or finding some other way to express the computation.

I hope that helps :)

**Comment by RunnersNum40** [comment]
No worries on the delay! That's a tricky issue, thank you for the clear explanation. I'll take comfort in knowing that my workaround is necessary in for this use case and not just a crazy hack.
